tempodb/backend/gcs/gcs.go

```diff
@@ -279,7 +278,15 @@ func (rw *readerWriter) readAll(ctx context.Context, name string) ([]byte, error
 	}
 	defer r.Close()
 
-	return ioutil.ReadAll(r)
+	// Preallocate the buffer with the exact size so we don't waste allocations
+	// while progressively growing an initial small buffer. The buffer capacity
+	// is increased by MinRead to avoid extra allocations due to how ReadFrom()
+	// internally works.
+	buf := bytes.NewBuffer(make([]byte, 0, r.Attrs.Size+bytes.MinRead))
+	if _, err := buf.ReadFrom(r); err != nil {
```

Comment:
Thoughts on using the new method pkg/io/ReadAllWithBuffer?  Would that avoid needing to take bytes.MinRead into account?

Reply:
yeah, i'd prefer `ReadAllWithBuffer()`. It will also cover cases where the size is incorrectly calculated for some reason. 

would prefer the same for all below.

Reply:
Cool will change. `ReadAllWithBuffer()` starts off with a 512 byte slice which is exactly what `bytes.MinRead` is

Reply:
If you pass in estimatedBytes that you can start with a buffer that will require far less allocations. You should be able to do that for most if not all of these changes.